File: samples/android/app/src/main/java/com/amazon/sampleapp/impl/Logger/LoggerHandler.java
/*
 * Copyright 2017-2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 *     http://aws.amazon.com/apache2.0/
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

package com.amazon.sampleapp.impl.Logger;

import android.graphics.Color;
import android.util.Log;

import com.amazon.aace.logger.Logger;
import com.amazon.sampleapp.core.LoggerControllerInterface;
import com.amazon.sampleapp.logView.LogEntry;
import com.amazon.sampleapp.logView.LogRecyclerViewAdapter;

import org.json.JSONException;
import org.json.JSONObject;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.PrintStream;
import java.text.SimpleDateFormat;
import java.util.Calendar;
import java.util.Date;
import java.util.LinkedList;
import java.util.Locale;
import java.util.Observable;
import java.util.Observer;
import java.util.Queue;

public class LoggerHandler extends Logger implements LoggerControllerInterface {
    /* Log colors */
    private static final int sColorVerbose = Color.parseColor("#B3E5FC"); // Light Blue
    private static final int sColorInfo = Color.parseColor("#FFFFFF"); // White
    private static final int sColorMetric = Color.parseColor("#73C54C"); // Light Green
    private static final int sColorWarn = Color.parseColor("#F57F17"); // Orange
    private static final int sColorError = Color.parseColor("#D50000"); // Red
    private static final int sColorJsonTemplate = Color.parseColor("#F9B702"); // Gold

    private static final SimpleDateFormat sTimeFormat // Note: not thread safe
            = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss.SSS", Locale.getDefault());
    private static final Level sLevel = Level.VERBOSE;
    private static final String sClientSourceTag = "CLI";

    private LoggerObservable mObservable;

    private Queue<LogEntry> mDisplayLogsQueue = new LinkedList<LogEntry>();
    private boolean mIsEnqueuingLogs = false;

    public LoggerHandler() {
        mObservable = new LoggerObservable();
    }

    // Handle log from Auto SDK
    @Override
    public boolean logEvent(Level level, long time, String source, String message) {
        if (level.ordinal() >= sLevel.ordinal()) {
            int color;
            switch (level) {
                case VERBOSE:
                    color = sColorVerbose;
                    break;
                case WARN:
                    color = sColorWarn;
                    break;
                case CRITICAL:
                case ERROR:
                    color = sColorError;
                    break;
                case INFO:
                    color = sColorInfo;
                    break;
                case METRIC:
                    color = sColorMetric;
                    break;
                default:
                    color = sColorInfo;
            }

            // Configure log for GUI log view
            JSONObject json = new JSONObject();
            try {
                json.put("text",
                        String.format("%-25s [%s] %3c%n%s", sTimeFormat.format(time), source, level.toChar(), message));
                json.put("textColor", color);
                json.put("source", source); // For log view filtering
                json.put("level", level.toString()); // For log view filtering
            } catch (JSONException e) {
                Log.e(sClientSourceTag, "Error: ", e);
                return true;
            }
            mObservable.log(json);
        }
        return true;
    }

    /* Client level log methods. Will use Auto SDK Logger */

    public void postVerbose(String tag, String message) {
        log(Level.VERBOSE, tag, message);
    }
    public void postInfo(String tag, String message) {
        log(Level.INFO, tag, message);
    }
    public void postWarn(String tag, String message) {
        log(Level.WARN, tag, message);
    }
    public void postError(String tag, String message) {
        log(Level.ERROR, tag, message);
    }
    public void postError(String tag, Throwable thr) {
        try (ByteArrayOutputStream os = new ByteArrayOutputStream()) {
            PrintStream ps = new PrintStream(os);
            thr.printStackTrace(ps);
            String str = os.toString();
            log(Level.ERROR, tag, str);

        } catch (IOException e) {
            Log.e(sClientSourceTag, "Error: ", e);
        }
    }

    /* Additional client logs. Will insert log into GUI log view but not use Auto SDK Logger */

    // Client log for JSON Templates
    public void postJSONTemplate(String tag, String message) {
        Date currentTime = Calendar.getInstance().getTime();
        Level level = Level.INFO;

        JSONObject json = new JSONObject();
        try {
            json.put("text",
                    String.format("%-25s [%s] %3c%n%s:%n%s", sTimeFormat.format(currentTime), sClientSourceTag,
                            level.toChar(), tag, message));
            json.put("textColor", sColorJsonTemplate);
            json.put("source", sClientSourceTag); // For log view filtering
            json.put("level", level.toString()); // For log view filtering

        } catch (JSONException e) {
            Log.e(sClientSourceTag, "Error: ", e);
            return;
        }

        synchronized (this) {
            if (mIsEnqueuingLogs) {
                mDisplayLogsQueue.add(new LogEntry(LogRecyclerViewAdapter.JSON_TEXT, json));
            } else {
                mObservable.log(json, LogRecyclerViewAdapter.JSON_TEXT);
            }
        }
    }

    // Client log for display cards
    public void postDisplayCard(JSONObject template, int logType) {
        Level level = Level.INFO;
        JSONObject json = new JSONObject();
        try {
            json.put("template", template);
            json.put("source", sClientSourceTag); // For log view filtering
            json.put("level", level.toString()); // For log view filtering

        } catch (JSONException e) {
            Log.e(sClientSourceTag, "Error: ", e);
        }

        synchronized (this) {
            if (mIsEnqueuingLogs) {
                mDisplayLogsQueue.add(new LogEntry(logType, json));
            } else {
                mObservable.log(json, logType);
            }
        }
    }

    public void pause() {
        synchronized (this) {
            mIsEnqueuingLogs = true;
        }
    }

    public void resume() {
        synchronized (this) {
            mIsEnqueuingLogs = false;

            while (!mDisplayLogsQueue.isEmpty()) {
                LogEntry entry = mDisplayLogsQueue.remove();
                int logType = entry.getType();
                JSONObject json = entry.getJSON();
                mObservable.log(json, logType);
            }
        }
    }

    /* Logger Observable for inserting logs into GUI log view */

    public static class LoggerObservable extends Observable {
        public void log(String message) {
            setChanged();
            notifyObservers(message);
        }

        public void log(JSONObject obj) {
            setChanged();
            notifyObservers(new LogEntry(LogRecyclerViewAdapter.TEXT_LOG, obj));
        }

        public void log(JSONObject obj, int logType) {
            setChanged();
            notifyObservers(new LogEntry(logType, obj));
        }
    }

    public void addLogObserver(Observer observer) {
        mObservable.addObserver(observer);
    }
}

File: samples/cpp/SampleApp/src/Logger/LoggerHandler.cpp
/*
 * Copyright 2018-2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 *     http://aws.amazon.com/apache2.0/
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

#include "SampleApp/Logger/LoggerHandler.h"

#include "SampleApp/ApplicationContext.h"

// C++ Standard Library
#include <ctime>
#include <iostream>
#include <regex>

// Guidelines Support Library
#define GSL_THROW_ON_CONTRACT_VIOLATION
#include <gsl/contracts.h>

namespace sampleApp {
namespace logger {

////////////////////////////////////////////////////////////////////////////////////////////////////
//
//  LoggerHandler
//
////////////////////////////////////////////////////////////////////////////////////////////////////

LoggerHandler::LoggerHandler(std::weak_ptr<Activity> activity) : m_activity{std::move(activity)} {
    setupUI();
}

std::weak_ptr<Activity> LoggerHandler::getActivity() {
    return m_activity;
}

// aace::logger::Logger interface

bool LoggerHandler::logEvent(
    Logger::Level level,
    std::chrono::system_clock::time_point time,
    const std::string& source,
    const std::string& message) {
    if (m_applicationContext->isLogEnabled() && (level >= m_applicationContext->getLevel())) {
        if (auto console = m_console.lock()) {
            std::string key{};
            switch (level) {
                case Level::VERBOSE:
                    key = "[V]";
                    break;
                case Level::INFO:
                    key = "[I]";
                    break;
                case Level::METRIC:
                    key = "[M]";
                    break;
                case Level::WARN:
                    key = "[W]";
                    break;
                case Level::ERROR:
                    key = "[E]";
                    break;
                case Level::CRITICAL:
                    key = "[C]";
                    break;
                default:
                    break;
            }
            if (!key.empty()) {
                console->printLine(source, key, message);
                return true;
            }
        }
    }
    return false;
}

// private

void LoggerHandler::setupUI() {
    auto activity = m_activity.lock();
    if (!activity) {
        return;
    }
    m_applicationContext = activity->getApplicationContext();
    m_console = activity->findViewById("id:console");

    // log
    activity->registerObserver(Event::onLoggerLog, [=](const std::string& value) {
        log(logger::LoggerHandler::Level::VERBOSE, "LoggerHandler", "onLoggerLog:" + value);
        static std::regex r("(.+)/(.+)/(.+)", std::regex::optimize);
        std::smatch sm{};
        if (!std::regex_match(value, sm, r) || ((sm.size() - 1) < 3)) {
            return false;
        }
        // clang-format off
        static const std::map<std::string, Level> LevelEnumerator{
            {"VERBOSE", Level::VERBOSE},
            {"INFO", Level::INFO},
            {"METRIC", Level::METRIC},
            {"WARN", Level::WARN},
            {"ERROR", Level::ERROR},
            {"CRITICAL", Level::CRITICAL}
        };
        // clang-format on
        if (LevelEnumerator.count(sm[1]) == 0) {
            return false;
        }
        log(LevelEnumerator.at(sm[1]), sm[2], sm[3]);
        return true;
    });
}

}  // namespace logger
}  // namespace sampleApp
